src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -316,40 +332,136 @@ public void insertBufferedRows() {
+   * channel is invalidated.
+   *
+   * <p>It can also send errors {@link
+   * net.snowflake.ingest.streaming.InsertValidationResponse.InsertError} in form of response inside
+   * {@link InsertValidationResponse}
+   *
+   * @param buffer buffer to insert into snowflake
+   * @return InsertValidationResponse object sent from insertRows API.
+   */
+  private InsertValidationResponse insertRowsWithFallback(StreamingBuffer buffer) {
+    Fallback<InsertValidationResponse> reopenChannelFallbackExecutorForInsertRows =
+        Fallback.builder(() -> insertRowsFallbackSupplier(buffer))
+            .handle(SFException.class)
+            .onFailure(
+                event ->
+                    LOGGER.error(
+                        "[INSERT_ROWS_FALLBACK] Failed to open Channel or insert Rows:{}",
+                        this.getChannelName(),
+                        event.getException()))
+            .build();
 
-    // should we retry if response.hasErrors() == true?
-    // how can we retry if buffer was cleared?
-    // lets have an exponential back off and jitter retry policy
+    return Failsafe.with(reopenChannelFallbackExecutorForInsertRows)
+        .get(new InsertRowsApiResponseSupplier(this.channel, buffer));
+  }
 
-    if (response.hasErrors()) {
-      LOGGER.error(
-          "Inserted rows API response has errors for rows with offset start:{} and end:{}",
-          intermediateBuffer.getFirstOffset(),
-          intermediateBuffer.getLastOffset());
-      response
-          .getInsertErrors()
-          .forEach(
-              insertError -> {
-                LOGGER.error(
-                    "Insert row Error message:{}, with ex:{}",
-                    insertError.getMessage(),
-                    insertError.getException().getMessage());
-              });
+  /** Invokes the API given the channel and streaming Buffer. */
+  private static class InsertRowsApiResponseSupplier
+      implements CheckedSupplier<InsertValidationResponse> {
+
+    private final SnowflakeStreamingIngestChannel channel;
+
+    private final StreamingBuffer streamingBuffer;
+
+    private InsertRowsApiResponseSupplier(
+        SnowflakeStreamingIngestChannel channel, StreamingBuffer streamingBuffer) {
+      this.channel = channel;
+      this.streamingBuffer = streamingBuffer;
+    }
+
+    @Override
+    public InsertValidationResponse get() throws Throwable {
+      LOGGER.debug(
+          "Invoking insertRows API for channel:{}, noOfRecords:{}, startOffset:{}, endOffset:{}",
+          this.channel.getFullyQualifiedName(),
+          streamingBuffer.getNumOfRecord(),
+          streamingBuffer.getFirstOffset(),
+          streamingBuffer.getLastOffset());
+      return this.channel.insertRows(
+          streamingBuffer.getData(), Long.toString(streamingBuffer.getLastOffset()));
+    }
+  }
+
+  /** We will reopen the channel on {@link SFException} and call insert rows API again. */
```

Comment:
We should reopen the channel and restart from the latest committed offset token, right?

Reply:
I suppose that is only for get offset token? 
Idea here is to retry sending the same set of rows and worry about resetting offset token only in getOffsetToken. resetting it in getoffsettoken makes sense because the rows will be sent again and hence insertRows can try inserting lost rows. 

Adding get offsetToken here just complicate things here since we would be sending rows for current offsets which might be higher than what we might reset later.

Reply:
If you follow [this](https://snowflake.slack.com/archives/G01F1B8PLDU/p1645582692556109) thread, it's possible that the queued chunks got skipped when we see a bigger client/channel sequencer, so you will need to fetch the latest offset token first

Reply:
Ahh okay. I kind of understand.. So calling insertRows here wouldnt make sense right?
Reopen and then reset.